core/primitives-core/src/serialize.rs

```diff
@@ -158,109 +158,136 @@ fn test_base_bytes_format() {
+        /// does not accept `null` values.
+        fn try_from_unit() -> Result<Self, ()> {
+            Err(())
+        }
+
+        /// Tries to parse decimal string as an integer.
+        fn try_from_str(value: &str) -> Result<Self, std::num::ParseIntError>;
+
+        /// Constructs Self from a 64-bit unsigned integer.
+        fn from_u64(value: u64) -> Self;
     }
 
-    pub fn deserialize<'de, D>(deserializer: D) -> Result<u64, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        let s = String::deserialize(deserializer)?;
-        u64::from_str_radix(&s, 10).map_err(de::Error::custom)
+    impl DecType for u64 {
+        fn serialize(&self) -> Option<String> {
+            Some(self.to_string())
+        }
+        fn try_from_str(value: &str) -> Result<Self, std::num::ParseIntError> {
+            Self::from_str_radix(value, 10)
+        }
+        fn from_u64(value: u64) -> Self {
+            value
+        }
     }
-}
 
-#[test]
-fn test_u64_dec_format() {
-    #[derive(PartialEq, Debug, serde::Deserialize, serde::Serialize)]
-    struct Test {
-        #[serde(with = "u64_dec_format")]
-        field: u64,
+    impl DecType for u128 {
+        fn serialize(&self) -> Option<String> {
+            Some(self.to_string())
+        }
+        fn try_from_str(value: &str) -> Result<Self, std::num::ParseIntError> {
+            Self::from_str_radix(value, 10)
+        }
+        fn from_u64(value: u64) -> Self {
+            value as u128
+        }
     }
 
-    assert_round_trip("{\"field\":\"42\"}", Test { field: 42 });
-    assert_round_trip("{\"field\":\"18446744073709551615\"}", Test { field: u64::MAX });
-    assert_de_error::<Test>("{\"field\":18446744073709551616}");
-    assert_de_error::<Test>("{\"field\":\"18446744073709551616\"}");
-    assert_de_error::<Test>("{\"field\":42}");
-}
+    impl<T: DecType> DecType for Option<T> {
+        fn serialize(&self) -> Option<String> {
+            self.as_ref().and_then(DecType::serialize)
+        }
+        fn try_from_unit() -> Result<Self, ()> {
+            Ok(None)
+        }
+        fn try_from_str(value: &str) -> Result<Self, std::num::ParseIntError> {
+            Some(T::try_from_str(value)).transpose()
+        }
+        fn from_u64(value: u64) -> Self {
+            Some(T::from_u64(value))
+        }
+    }
 
-pub mod u128_dec_format {
-    use serde::de;
-    use serde::{Deserialize, Deserializer, Serializer};
+    struct Visitor<T>(core::marker::PhantomData<T>);
 
-    pub fn serialize<S>(num: &u128, serializer: S) -> Result<S::Ok, S::Error>
+    impl<'de, T: DecType> de::Visitor<'de> for Visitor<T> {
+        type Value = T;
+
+        fn expecting(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+            fmt.write_str("a non-negative integer as a string")
+        }
+
+        fn visit_unit<E: de::Error>(self) -> Result<T, E> {
+            T::try_from_unit().map_err(|_| de::Error::invalid_type(de::Unexpected::Option, &self))
```

Comment:
```suggestion
            T::try_from_unit().map_err(|()| de::Error::invalid_type(de::Unexpected::Option, &self))
```

Makes it very slightly easier to audit for `map_err`s that discard error causes if we ever do it in the future. Don’t have strong feelings either way though, so if you’d rather not, I’m okay with leaving as-is.